cpp/src/arima/batched_arima.cu

```diff
@@ -0,0 +1,259 @@
+                         d_y_p[it] = d_y[it + 1] - d_vs[itd];
+                         d_y_diff[itd] = d_y[it + 1] - d_y[it];
+                       }
+                     });
+  }
+
+  // due to `differencing` we need to forecast a single step to make the
+  // in-sample prediction the same length as the original signal.
+  if (d == 1) {
+    double* d_y_fc = (double*)handle.getDeviceAllocator()->allocate(
+      sizeof(double) * num_batches, handle.getStream());
+    forecast(handle, 1, p, d, q, num_batches, nobs, d_y, d_y_diff, d_vs,
+             d_params, d_y_fc);
+
+    // append forecast to end of in-sample prediction
+    auto counting = thrust::make_counting_iterator(0);
+    thrust::for_each(thrust::cuda::par.on(stream), counting,
+                     counting + num_batches, [=] __device__(int bid) {
+                       d_y_p[bid * nobs + (nobs - 1)] = d_y_fc[bid];
+                     });
+  }
+}
+
+void batched_loglike(cumlHandle& handle, double* d_y, int num_batches, int nobs,
+                     int p, int d, int q, double* d_params,
+                     std::vector<double>& loglike, double* d_vs, bool trans) {
+  using std::get;
+  using std::vector;
+
+  ML::PUSH_RANGE(__FUNCTION__);
+
+  // unpack parameters
+  auto allocator = handle.getDeviceAllocator();
+  auto stream = handle.getStream();
+  double* d_mu =
+    (double*)allocator->allocate(sizeof(double) * num_batches, stream);
+  double* d_ar =
+    (double*)allocator->allocate(sizeof(double) * num_batches * p, stream);
+  double* d_ma =
+    (double*)allocator->allocate(sizeof(double) * num_batches * q, stream);
+  double* d_Tar =
+    (double*)allocator->allocate(sizeof(double) * num_batches * p, stream);
+  double* d_Tma =
+    (double*)allocator->allocate(sizeof(double) * num_batches * q, stream);
+
+  // params -> (mu, ar, ma)
+  unpack(d_params, d_mu, d_ar, d_ma, num_batches, p, d, q, handle.getStream());
+
+  CUDA_CHECK(cudaPeekAtLastError());
+
+  if (trans) {
+    batched_jones_transform(handle, p, q, num_batches, false, d_ar, d_ma, d_Tar,
+                            d_Tma);
+  } else {
+    // non-transformed case: just use original parameters
+    CUDA_CHECK(cudaMemcpyAsync(d_Tar, d_ar, sizeof(double) * num_batches * p,
+                               cudaMemcpyDeviceToDevice, stream));
+    CUDA_CHECK(cudaMemcpyAsync(d_Tma, d_ma, sizeof(double) * num_batches * q,
+                               cudaMemcpyDeviceToDevice, stream));
+  }
+
+  if (d == 0) {
+    // no diff
+    batched_kalman_filter(handle, d_y, nobs, d_Tar, d_Tma, p, q, num_batches,
+                          loglike, d_vs);
+  } else if (d == 1) {
+    ////////////////////////////////////////////////////////////
+    // diff and center (with `mu`):
+    ////////////////////////////////////////////////////////////
+
+    // make device array and pointer
+    double* y_diff = (double*)allocator->allocate(
+      num_batches * (nobs - 1) * sizeof(double), stream);
```

Comment:
This one is never deallocated, which causes an out of memory error on large datasets.

You can add before the `} else {` line:
```c++
allocator->deallocate(y_diff, sizeof(double) * num_batches * (nobs - 1),
                      stream);
```

Reply:
Done.